models/staging/ga4/stg_ga4__user_id_mapping.sql

```diff
@@ -0,0 +1,29 @@
+with events_with_user_id as (
+    select 
+        user_id,
+        user_pseudo_id,
+        event_timestamp 
+    from {{ref('stg_ga4__events')}}
+    where user_id is not null
+        and user_pseudo_id is not null
+),
+include_last_seen_timestamp as (
+    select 
+        user_id,
+        user_pseudo_id,
+        max(event_timestamp) as last_seen_timestamp
+    from events_with_user_id
+    group by 1,2
+),
+pick_latest_timestamp as (
```

Comment:
This might be confusion on my part as I don't know these changes as well as you and it depends on how the `last_seen_timestamp` is used in subsequent models.

It is possible, though, that what you want for `last_seen_timestamp` is the last timestamp mapped to a `user_pseudo_id` regardless of the presence or absence of a `user_id`. This is to account for  sessions that extend after logging out.

However, it is possible that our session models are sufficient to account for these cases and that we want to preserve this as `user_id` focused. We'll need to communicate this.

Reply:
Currently, the `last_seen_timestamp` field is only used within this model to pick a winner when multiple user_ids may have been associated with a single `user_pseudo_id` (which should be rare, but possible. Example would be a library computer where a single browser might see different users log in). 

So maybe `last_seen` is too vague and could be renamed to `last_seen_user_id_timestamp` to make it clear this isn't the last time we've seen the user's device. 

So going back to your example of a logout... the user's events after logout would be associated with their `user_id` in the `fct_user_ids` model which I think is the expected behavior.